general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+      {
+         // median is at itb+siz/2
+         level=level+1;
+         if ((pt.xx[dim]-R)>mtb->pt.xx[dim]) // look to the right only
+         {
+            RNS(pt, R, itb+siz/2+1, ite, level, res, dist);
+         }
+         else if ((pt.xx[dim]+R)<mtb->pt.xx[dim]) // look to the left only
+         {
+            RNS(pt,R, itb, itb+siz/2, level,   res, dist);
+         }
+         else  // check all
+         {
+            RNS(pt,R, itb+siz/2+1, ite, level, res, dist); // right
+            RNS(pt,R, itb, itb+siz/2, level,   res, dist); // left
+
+            // check central one
+            Tfloat dd=Dist(mtb->pt, pt);
+            if (dd<R)
+            {
+               res.push_back(mtb->ind);
+               dist.push_back(dd);
+            }
+         }
+      }
+      else
+      {
+         Tfloat dd;
+         for (auto it=itb; it!=ite; it++)
+         {
+            dd=Dist(it->pt, pt);
+            if (dd<R) // update bc
+            {
+               res.push_back(it->ind);
+               dist.push_back(dd);
+            }
+         }
+      }
+   }
+
+
+};
+
+
+/// Defines KDTree in 3D
+typedef KDTree<int,double,3> KDTree3D;
+
+/// Defines Contructs KDTree in 2D
+typedef KDTree<int,double,2> KDTree2D;
+
+/// Defines KDTree in 1D
+typedef KDTree<int,double,1> KDTree1D;
+
+
+/// Defines KDTree with points stored outside the class.
+/// The external storage slows down the search and the construction
+/// methods.
+class KDTreeM
+{
+public:
+   /// Construtor
+   KDTreeM(mfem::Vector& coords_,int dim_)
+   {
+      gf=&coords_;
+      ndim=dim_;
+      tp.SetSize(ndim);
+      int np=(gf->Size())/ndim;
+      data.resize(np);
+      for (int i=0; i<np; i++)
+      {
+         data[i].xx=coords_.GetData()+i*ndim;
+         data[i].ind=i;
+      }
+      Sort();
+   }
+
+   /// Internal structure used as a node in the KDTree
+   struct NodeND
+   {
+      size_t ind;
+      double *xx;//coordinates of the node
+   };
+
+
+   /// Returns the spatial dimension of the points
+   int SpatialDimension() {return ndim;}
+
+   /// Data iterator
+   typedef typename std::vector<NodeND>::iterator iterator;
+
+   /// Returns iterator to beginning of the point cloud
+   iterator begin() {return data.begin();}
+
+   /// Returns iterator to the end of the point cloud
+   iterator end() {return data.end();}
+
+   /// Returns the size of the point cloud
+   size_t size() {return data.size();}
+
+   /// Clears the point cloud.
+   void clear() { data.clear();}
+
+   /// Builds the KDTree.
+   void Sort() { SortInPlace(data.begin(),data.end(),0);}
+
+   /// Returns the index of the closest point to the input point pt.
+   int FindClosestPoint(mfem::Vector& pt);
+
+   ///  Returns the closest point and the distance to the input point pt.
+   void FindClosestPoint(Vector& pt, int &ind, double& dist);
+
+   /// Brute force search - please, use it only for debuging purposes.
+   void _FindClosestPoint(Vector& pt,int& ind, double& dist);
```

Comment:
It works on the particular data structure - It can be defined as independent functions which will take the data structure as argument, however, the implementation will be duplicated.

Reply:
As a rule of thump I learned that, if an algorithm acts on a data structure, then it should be a free function and if it maintains data invariants, then it should be a member function.

I also just realized that the distance type does not match the return type of the norm.

Reply:
To quickly elaborate on this rule: It helps to modularize and generalize the code, and respects to the open-closed principle.

I think I also miss the point why the implementation will be duplicated, because we should be able to write the algorithm in a generic way. Could you quickly comment on this?

Reply:
I will remove this class - it is much slower than the templated one and I do not see good applications right now.